src/compiler/visitor.ts

```diff
@@ -0,0 +1,830 @@
+                    result = isArray(value)
+                        ? reduceLeft(<NodeArray<Node>>value, f, result)
+                        : f(result, <Node>value);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Visits a Node using the supplied visitor, possibly returning a new Node in its place.
+     *
+     * @param node The Node to visit.
+     * @param visitor The callback used to visit the Node.
+     * @param test A callback to execute to verify the Node is valid.
+     * @param optional An optional value indicating whether the Node is itself optional.
+     * @param lift An optional callback to execute to lift a NodeArrayNode into a valid Node.
+     */
+    export function visitNode<T extends Node>(node: T, visitor: (node: Node) => Node, test: (node: Node) => boolean, optional?: boolean, lift?: (node: NodeArray<Node>) => T): T {
+        if (node === undefined) {
+            return undefined;
+        }
+
+        const visited = visitor(node);
+        if (visited === node) {
+            return node;
+        }
+
+        const lifted = liftNode(visited, lift);
+        if (lifted === undefined) {
+            Debug.assert(optional, "Node not optional.");
+            return undefined;
+        }
+
+        Debug.assert(test === undefined || test(visited), "Wrong node type after visit.");
+        return <T>visited;
+    }
+
+    /**
+     * Visits a NodeArray using the supplied visitor, possibly returning a new NodeArray in its place.
+     *
+     * @param nodes The NodeArray to visit.
+     * @param visitor The callback used to visit a Node.
+     * @param test A node test to execute for each node.
+     * @param start An optional value indicating the starting offset at which to start visiting.
+     * @param count An optional value indicating the maximum number of nodes to visit.
+     */
+    export function visitNodes<T extends Node, TArray extends NodeArray<T>>(nodes: TArray, visitor: (node: Node) => Node, test: (node: Node) => boolean, start?: number, count?: number): TArray {
+        if (nodes === undefined) {
+            return undefined;
+        }
+
+        let updated: T[];
+
+        // Ensure start and count have valid values
+        const length = nodes.length;
+        if (start === undefined || start < 0) {
+            start = 0;
+        }
+
+        if (count === undefined || count > length - start) {
+            count = length - start;
+        }
+
+        // If we are not visiting all of the original nodes, we must always create a new array.
+        if (start > 0 || count < length) {
+            updated = [];
+        }
+
+        // Visit each original node.
+        for (let i = 0; i < count; i++) {
+            const node = nodes[i + start];
+            const visited = node && <OneOrMore<T>>visitor(node);
+            if (updated !== undefined || visited === undefined || visited !== node) {
+                if (updated === undefined) {

```

Comment:
If you visit a NodeArray with `visitNodes`, you want to return a reference to the original `nodes` array if there are no changes. This is to allow us to use a less expensive reference equality test to check for differences when visiting the tree, rather than a more expensive check of each node in the resulting array.

When the `visitor` callback returns `undefined`, it signals that the visited node has no corresponding node in the output. As a result, if we have not yet allocated the `updated` node array, we need to make sure that it is allocated. Otherwise, if you were to visit a node array with three nodes, and all of them were to be removed, we would instead return the original node array.